tests/connection_state_test.c

```diff
@@ -0,0 +1,947 @@
+}
+
+static bool s_is_connection_completed(void *arg) {
+    struct mqtt_connection_state_test *state_test_data = arg;
+    return state_test_data->connection_completed;
+}
+
+static void s_wait_for_connection_to_complete(struct mqtt_connection_state_test *state_test_data) {
+    aws_mutex_lock(&state_test_data->lock);
+    aws_condition_variable_wait_pred(
+        &state_test_data->cvar, &state_test_data->lock, s_is_connection_completed, state_test_data);
+    aws_mutex_unlock(&state_test_data->lock);
+}
+
+void s_on_disconnect_fn(struct aws_mqtt_client_connection *connection, void *userdata) {
+    struct mqtt_connection_state_test *state_test_data = userdata;
+    aws_mutex_lock(&state_test_data->lock);
+    state_test_data->client_disconnect_completed = true;
+    aws_mutex_unlock(&state_test_data->lock);
+
+    aws_condition_variable_notify_one(&state_test_data->cvar);
+}
+
+static bool s_is_disconnect_completed(void *arg) {
+    struct mqtt_connection_state_test *state_test_data = arg;
+    return state_test_data->client_disconnect_completed && state_test_data->server_disconnect_completed;
+}
+
+static void s_wait_for_disconnect_to_complete(struct mqtt_connection_state_test *state_test_data) {
+    aws_mutex_lock(&state_test_data->lock);
+    aws_condition_variable_wait_pred(
+        &state_test_data->cvar, &state_test_data->lock, s_is_disconnect_completed, state_test_data);
+    state_test_data->client_disconnect_completed = false;
+    state_test_data->server_disconnect_completed = false;
+    aws_mutex_unlock(&state_test_data->lock);
+}
+
+static void s_on_publish_received(
+    struct aws_mqtt_client_connection *connection,
+    const struct aws_byte_cursor *topic,
+    const struct aws_byte_cursor *payload,
+    void *userdata) {
+    (void)connection;
+    (void)topic;
+    struct mqtt_connection_state_test *state_test_data = userdata;
+
+    struct aws_byte_buf payload_cp;
+    aws_byte_buf_init_copy_from_cursor(&payload_cp, state_test_data->allocator, *payload);
+
+    aws_mutex_lock(&state_test_data->lock);
+    aws_array_list_push_back(&state_test_data->published_messages, &payload_cp);
+    state_test_data->publishes_received++;
+    aws_mutex_unlock(&state_test_data->lock);
+    aws_condition_variable_notify_one(&state_test_data->cvar);
+}
+
+static bool s_is_publish_received(void *arg) {
+    struct mqtt_connection_state_test *state_test_data = arg;
+    return state_test_data->publishes_received == state_test_data->expected_publishes;
+}
+
+static void s_wait_for_publish(struct mqtt_connection_state_test *state_test_data) {
+    aws_mutex_lock(&state_test_data->lock);
+    aws_condition_variable_wait_pred(
+        &state_test_data->cvar, &state_test_data->lock, s_is_publish_received, state_test_data);
+    state_test_data->publishes_received = 0;
+    state_test_data->expected_publishes = 0;
+    aws_mutex_unlock(&state_test_data->lock);
+}
+
+static void s_on_suback(
```

Comment:
Should we also track unsubacks? Probably right?

Reply:
we're not really testing the server.

Reply:
so the act of the test not hanging tests that the client does its job.